go/store/datas/pull/pull_chunk_tracker_test.go

```diff
@@ -0,0 +1,169 @@
+)
+
+func TestPullChunkTracker(t *testing.T) {
+	t.Run("Empty", func(t *testing.T) {
+		tracker := NewPullChunkTracker(context.Background(), make(hash.HashSet), TrackerConfig{
+			BatchSize:          64 * 1024,
+			HasManyThreadCount: 3,
+			Haser:              nil,
+		})
+		hs, ok, err := tracker.GetChunksToFetch()
+		assert.Len(t, hs, 0)
+		assert.False(t, ok)
+		assert.NoError(t, err)
+		tracker.Close()
+	})
+
+	t.Run("HasAllInitial", func(t *testing.T) {
+		hs := make(hash.HashSet)
+		for i := byte(0); i < byte(10); i++ {
+			var h hash.Hash
+			h[0] = i
+			hs.Insert(h)
+		}
+		tracker := NewPullChunkTracker(context.Background(), hs, TrackerConfig{
+			BatchSize:          64 * 1024,
+			HasManyThreadCount: 3,
+			Haser:              hasAllHaser{},
+		})
+		hs, ok, err := tracker.GetChunksToFetch()
+		assert.Len(t, hs, 0)
+		assert.False(t, ok)
+		assert.NoError(t, err)
+		tracker.Close()
+	})
+
+	t.Run("HasNoneInitial", func(t *testing.T) {
+		hs := make(hash.HashSet)
+		for i := byte(1); i <= byte(10); i++ {
+			var h hash.Hash
+			h[0] = i
+			hs.Insert(h)
+		}
+		tracker := NewPullChunkTracker(context.Background(), hs, TrackerConfig{
+			BatchSize:          64 * 1024,
+			HasManyThreadCount: 3,
+			Haser:              hasNoneHaser{},
+		})
+		hs, ok, err := tracker.GetChunksToFetch()
+		assert.Len(t, hs, 10)
+		assert.True(t, ok)
+		assert.NoError(t, err)
+		hs, ok, err = tracker.GetChunksToFetch()
+		assert.Len(t, hs, 0)
+		assert.False(t, ok)
+		assert.NoError(t, err)
+
+		for i := byte(1); i <= byte(10); i++ {
+			var h hash.Hash
+			h[1] = i
+			tracker.Seen(h)
+		}
+
+		cnt := 0
+		for {
+			hs, ok, err := tracker.GetChunksToFetch()
+			assert.NoError(t, err)
+			if !ok {
+				assert.Equal(t, 10, cnt)
+				break
+			}
+			cnt += len(hs)
+		}
+
+		tracker.Close()
+	})
+
+	t.Run("HasManyError", func(t *testing.T) {
+		hs := make(hash.HashSet)
+		for i := byte(0); i < byte(10); i++ {
+			var h hash.Hash
+			h[0] = i
+			hs.Insert(h)
+		}
+		tracker := NewPullChunkTracker(context.Background(), hs, TrackerConfig{
+			BatchSize:          64 * 1024,
+			HasManyThreadCount: 3,
+			Haser:              errHaser{},
+		})
+		_, _, err := tracker.GetChunksToFetch()
+		assert.Error(t, err)
+		tracker.Close()
+	})
+
+	t.Run("InitialAreSeen", func(t *testing.T) {
+		hs := make(hash.HashSet)
+		for i := byte(0); i < byte(10); i++ {
+			var h hash.Hash
+			h[0] = i
+			hs.Insert(h)
+		}
+		tracker := NewPullChunkTracker(context.Background(), hs, TrackerConfig{
+			BatchSize:          64 * 1024,
+			HasManyThreadCount: 3,
+			Haser:              hasNoneHaser{},
+		})
+		hs, ok, err := tracker.GetChunksToFetch()
+		assert.Len(t, hs, 10)
+		assert.True(t, ok)
+		assert.NoError(t, err)
+
+		for i := byte(0); i < byte(10); i++ {
+			var h hash.Hash
+			h[0] = i
+			tracker.Seen(h)
+		}
+
+		hs, ok, err = tracker.GetChunksToFetch()
+		assert.Len(t, hs, 0)
+		assert.False(t, ok)
+		assert.NoError(t, err)
+
+		tracker.Close()
+	})
+}
```

Comment:
I don't see any tests where there are mixed results - ie have some and not others. Seems like that will be a common situation and there should be a test for it.

Reply:
Done. Thanks!